### PR TITLE
Add allocatePartyWithHint(On) to DAML Script

### LINKED ISSUE #4472: Add `allocatePartyWithHint` and `allocatePartyWithHintOn` to DAML script
We need variants of `allocateParty` that allow you to supply a party id hint. This is particularly useful on sandbox since it will use the hint as the exact party id so it allows you to recover the behavior from scenarios’ `getParty` function.

cc @aherrmann

### PR DESCRIPTION
**Pull request by aherrmann-da**
Closes #4472 

- Adds `allocatePartyWithHint` and `allocatePartyWithHintOn` to specify the `party_id_hint` from DAML script.
- Wraps the `PartyIdHint` and `ParticipantName` arguments in `newtype` wrappers to avoid confusing and error-prone types such as `allocatePartyWithHintOn : Text -> Text -> Text -> Script Party`.
- Adds test cases for the single and multi participant cases.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-script/daml/Daml/Script.daml [hunk 1]
```diff
@@ -92,6 +96,7 @@ query p = Script $ Free $ Query (QueryACS p (templateTypeRep @t) (pure . map (\(
 
 data AllocateParty a = AllocateParty
   { displayName : Text
+  , idHint : Optional Text
```
Comment (cocreature) [inline daml-script/daml/Daml/Script.daml]: Protobuf and therefore the ledger API doesn’t differentiate between an empty string and a string not being present. You can therefore just remove the `Optional` here.

File: daml-script/daml/Daml/Script.daml [hunk 2]
```diff
@@ -101,15 +106,51 @@ data SleepRec a = SleepRec
...
+-- Must be a valid PartyIdString (as described in @value.proto@).
+newtype PartyIdHint = PartyIdHint { partyIdHint : Text }
+
+newtype ParticipantName = ParticipantName { participantName : Text }
+
 -- | Allocate a party with the given display name
 -- using the party management service.
 allocateParty : Text -> Script Party
-allocateParty displayName = Script $ Free (AllocParty $ AllocateParty displayName None pure)
+allocateParty displayName = Script $ Free (AllocParty $ AllocateParty
```
Comment (cocreature) [inline daml-script/daml/Daml/Script.daml]:
```suggestion
allocateParty displayName = Script $ Free $ AllocParty AllocateParty
```

File: daml-script/daml/Daml/Script.daml [hunk 3]
```diff
@@ -101,15 +106,51 @@ data SleepRec a = SleepRec
...
+  { displayName
+  , idHint = None
+  , participant = None
+  , continue = pure
+  })
```
Comment (cocreature) [inline daml-script/daml/Daml/Script.daml]:
```suggestion
  }
```

File: daml-script/daml/Daml/Script.daml (hunk 1 above)
Comment (aherrmann-da) [reply daml-script/daml/Daml/Script.daml]: Indeed, done.

File: daml-script/daml/Daml/Script.daml (hunk 2 above)
Comment (aherrmann-da) [reply daml-script/daml/Daml/Script.daml]: done

File: daml-script/daml/Daml/Script.daml (hunk 3 above)
Comment (aherrmann-da) [reply daml-script/daml/Daml/Script.daml]: done

**Pull request merged**